pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -419,6 +421,42 @@ func (c *Controller) deleteService(svc *model.Service) {
 	c.handlers.NotifyServiceHandlers(nil, svc, event)
 }
 
+// recomputeServiceForPod is called when a pod changes and service endpoints need to be recompute
+// Most of Pod is immutable, so once it has been created we are ok to cache the internal representation.
+// However, a few fields (labels) are mutable. When these change, we call recomputeServiceForPod and rebuild the cache
+// for all service's the pod is a part of and push an update.
+func (c *Controller) recomputeServiceForPod(pod *v1.Pod) {
+	allServices := c.services.List(pod.Namespace, klabels.Everything())
+	cu := sets.New[model.ConfigKey]()
+	services := getPodServices(allServices, pod)
```

Comment:
`getPodServices` is very costy, especially when services number is huge.

Reply:
This function is called extremely rarely -- only if a pod's labels change. We already call this WAY more often in other codepaths (GetProxyServiceTargets). I don't think its so bad to list a namespace's services every time a pod changes.. even a large cluster isn't going to have >1k services in one namespace probably?

The complexity to optimize this is not worth it IMO -- it would be very complex

Reply:
> even a large cluster isn't going to have >1k services in one namespace probably?

Maybe

>This function is called extremely rarely

Suggest we add a feature flag, this could be a breaking change.

Reply:
> Suggest we add a feature flag, this could be a breaking change.

How can it be a breaking change? Or just  a performance regression you mean?

Reply:
Performance regression i mean